[Infrastructure] Use MSBuild v15 extension point to control CL parallelism, abstract MSBuild /maxcpucount switch

4creators opened the pull request:
Fixes #14543
Fixes #14605

This commit frees -ExtraParameters and uses MSBuild v15 extension point to control CL compiler parallelism.

1. If MSBuild 15.0 is available - installed with Visual Studio 2017 - CL parallelism control is achieved by using extension point in `Microsoft.Common.props` file which allows to include `Directory.Build.props` - for details see [Microsoft.Common.props] [1] file in MSBuild repo. https://github.com/Microsoft/msbuild/blob/b38e4ceeaaec36c5237ae698041e9b9f18c84876/src/Tasks/Microsoft.Common.props#L36. 

2. MSBuild parallelism is controled via abstracted /maxcpucount command line switch
  which now is available as -MSBuildNodeCount pass through run.cmd/run.sh command line
  argument with default value equal to /maxcpucount (what defaults to number of logical processors).
  This allows to control MSBuild parallelism on all supported platforms.

It is possible to control MSBuild parallelism programatically via MSBuild public
API using MaxNodeCount properties available on the following APIs:

- Microsoft.Build.Evaluation.ProjectCollection.MaxNodeCount
- Microsoft.Build.Execution.BuildParameters.MaxNodeCount

It's use in current build infrastructure would require creating unnecessary
complexity and was ruled out.

4creators commented:
PTAL
@BruceForstall @weshaggard @danmosemsft 

Changes required to move control of CL parallelism to CMake script forced unification of the way adding of CMake targets is done. Now only functions from functions.cmake are used which inject required parameter to vcxprojects. At this same time it was necessary to increase min. CMake version on Windows to v3.5. It is not necessary on Unixes but seems desirable. I do not know what effect this will have on CI since I don't now CI machines configuration.

4creators commented:
Dotnet bot test please OSX10.12 x64 Checked Build and Test

4creators commented:
@dotnet-bot test please OSX10.12 x64 Checked Build and Test

4creators commented:
@dotnet-bot help

janvorli commented on functions.cmake:
```diff
@@ -216,6 +216,11 @@ function(_add_executable)
       disable_pax_mprotect(${ARGV})
     else()
       add_executable(${ARGV})
+      list(GET ARGV 0 CLR_CMAKE_CURRENT_TARGET)
```
This extraction is not needed, ARGV0 variable can be used to access  that argument. See the usage in the code below.

4creators replied:
Will fix. Thanks

janvorli commented on functions.cmake:
```diff
@@ -216,6 +216,11 @@ function(_add_executable)
+      set_target_properties(
+        ${CLR_CMAKE_CURRENT_TARGET} 
+        PROPERTIES
+        VS_GLOBAL_CL_MPCount ${CLR_CMAKE_CPU_CORES})
```
A nit - coding convention - the closing ) should be on a separate line aligned with the set_target_properties.

4creators replied:
Will fix

janvorli commented on functions.cmake:
```diff
@@ -228,13 +233,31 @@ function(_add_library)
       add_library(${ARGV} ${VERSION_FILE_PATH})
     else()
       add_library(${ARGV})
+      list(GET ARGV 0 CLR_CMAKE_CURRENT_TARGET)
```
The same as above

janvorli commented on functions.cmake:
```diff
@@ -228,13 +233,31 @@ function(_add_library)
     endif()  
 endfunction()
 
+function(_add_custom_target)
```
Why does the custom target need this option? Custom targets don't build anything themselves, they are just kind of proxies to related custom_command that actually does some work.

4creators replied:
I need to inject <CL_MPCount>x</CL_MPCount> property into vcxproj which actually builds code. It just happens that mechanism to perform that injection with VS_GLOBAL_<variable> feature works only on targets which translate directly to vcxproj on Windows. Custom targets do that but custom commands not - it is just CMake implementation detail which has surfaced in public API.

janvorli replied:
Ok, that would make sense, but as for the single use of this new _add_custom_target that you have added, the mscordaccore_exp doesn't result in CL invocation, it invokes lib.exe instead. Does the option still apply?

janvorli replied:
I agree that centralizing the change is a good thing. So then it seems we should change all add_custom_target usages by the new _add_custom_target. There are 16 usages of that.

4creators replied:
I will do that then

janvorli commented on functions.cmake:
```diff
@@ -228,13 +233,31 @@ function(_add_library)
+      add_custom_target(${ARGV})
+    else()
+      add_custom_target(${ARGV})
+      list(GET ARGV 0 CLR_CMAKE_CURRENT_TARGET)
```
Here too, this extraction is not needed, ARGV0 variable can be used to access that argument. See the usage in the code below.

4creators replied:
Will fix

4creators commented:
DO-NOT-MERGE until notified.

@janvorli @weshaggard 

I am working on fixing build errors for Windows cross target builds -> **arm64 Linux** which fail to build some binaries due to CMake excluding their projects from build. This seems to be a reason for other failures as well as the jobs failing are subjobs run on Windows.

Since I have no tools to run it on my machine I am struggling to get better info from CI builds. I am going to start tracing cmake runs on CI machines, so even if all builds pass DO-NOT-MERGE until notified.

Any help would be appreciated.

janvorli commented:
@4creators I'll take a look at the logs once the CI legs fail. But we don't cross build for arm64 Linux on Windows.

4creators commented:
I was not precise - it is not exactly cross build to Linux but to linuxmscorlib:
```
D:\j\workspace\x64_release_w---0575cb46>build.cmd release arm64 linuxmscorlib
```

The error is caused by this:
```
BUILD: Commencing build of cross architecture native components for Linux.arm64.Release
BUILD: Using environment: "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\Tools\\..\..\VC\Auxiliary\Build\vcvarsall.bat" x86_amd64
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0.26730.16
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x86_x64'
-- The C compiler identification is MSVC 19.11.25508.2
-- The CXX compiler identification is MSVC 19.11.25508.2
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.11.25503/bin/Hostx86/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.11.25503/bin/Hostx86/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.11.25503/bin/Hostx86/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.11.25503/bin/Hostx86/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- VS_PLATFORM_TOOLSET is v141
-- VS_PLATFORM_NAME is x64
-- The ASM_MASM compiler identification is MSVC
-- Found assembler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.11.25503/bin/Hostx86/x64/ml64.exe
-- Number of processor cores: 4
-- Configuring done
WARNING: Target "superpmi" has EXCLUDE_FROM_ALL set and will not be built by default but an install rule has been provided for it.  CMake does not define behavior for this case.
WARNING: Target "mcs" has EXCLUDE_FROM_ALL set and will not be built by default but an install rule has been provided for it.  CMake does not define behavior for this case.
WARNING: Target "superpmi-shim-collector" has EXCLUDE_FROM_ALL set and will not be built by default but an install rule has been provided for it.  CMake does not define behavior for this case.
WARNING: Target "superpmi-shim-collector" has EXCLUDE_FROM_ALL set and will not be built by default but an install rule has been provided for it.  CMake does not define behavior for this case.
WARNING: Target "superpmi-shim-counter" has EXCLUDE_FROM_ALL set and will not be built by default but an install rule has been provided for it.  CMake does not define behavior for this case.
WARNING: Target "superpmi-shim-counter" has EXCLUDE_FROM_ALL set and will not be built by default but an install rule has been provided for it.  CMake does not define behavior for this case.
WARNING: Target "superpmi-shim-simple" has EXCLUDE_FROM_ALL set and will not be built by default but an install rule has been provided for it.  CMake does not define behavior for this case.
WARNING: Target "superpmi-shim-simple" has EXCLUDE_FROM_ALL set and will not be built by default but an install rule has been provided for it.  CMake does not define behavior for this case.
```
I have no idea why EXCLUDE_FROM_ALL is set by CMake for above projects.

4creators commented:
@janvorli 

Seems that the reason was in additional logic through which targets were created included test if required corr component was on list and previously not checked projects were never on the list however they were built and included during cross component build.

janvorli commented:
@BruceForstall - do the superpmi binaries belong to cross components, that means components that we build to run on one architecture but working with code of another one (like crossgen that can run on amd64 and generate code for arm64)?

4creators commented:
Looks like CI is failing due to infrastructure problems

janvorli commented:
It is interesting that it fails right after building superpmi. It seems it could be related to your changes.

4creators commented:
There are some IO problems. I will switch off CMake tracing and try again.

BruceForstall commented:
> @BruceForstall - do the superpmi binaries belong to cross components, that means components that we build to run on one architecture but working with code of another one (like crossgen that can run on amd64 and generate code for arm64)?

I've never tested it (or thought about it) in that mode, so I don't think it should be a cross component.

4creators commented:
OK I will remove them from  cross component windows targets

weshaggard commented on functions.cmake:
```diff
@@ -228,13 +233,31 @@ function(_add_library)
+      add_custom_target(${ARGV})
+      set_target_properties(
+        ${ARGV0} 
+        PROPERTIES
```
My Cmake skills are some what limited but it would seem like there should be a way to add this property globally instead of requiring this wrapper. Is that not true?

weshaggard replied:
Ideally something like if WIN32 set this global property.

4creators replied:
The problem is that it is so called target property so it can only be set for a given target. I have tried to set it as global property but it had no effect on generated projects. It specifically defined as a property which can be set on targets in CMake documentation:

https://cmake.org/cmake/help/v3.9/manual/cmake-properties.7.html

janvorli replied:
There is nothing like that for target properties like this one. However, googling a bit out of interest has revealed there was something that can help added to CMake in version 3.7 -  BUILDSYSTEM_TARGETS property that contains a list of all targets. So if we moved to this version of cmake as the minimum for Windows, we could use that at the end of the main CMakeFile.txt to set the property for all targets.

4creators replied:
In general CMake is not very expressive in Visual Studio generators and perhaps it would necessary to submit to their repo a PR with general feature enabling custom modifications of vcxprojects like:
- VS_GLOBAL_NODE_<variable> -> accepts complex values
- with value set to arguments of EmitNode generator function arguments
- EmitNode(string & nodeName, string* attributeNames, string* attributeValues, string & nodeValue, string & parentNodeXPath);

4creators replied:
I have made some preliminary tests on CMake 3.9 with BUILDSYSTEM_TARGETS before going with this solution and it's value was always empty at the top level scope. Additional test inside _add_library call which added target were negative as well - BUILDSYSTEM_TARGETS was empty despite that target existed since it was added one code line earlier. I have not investigated it any further.

janvorli replied:
Please note it is not a variable, but a directory property. Maybe you know and read it that way, but just in case you didn't ...

4creators replied:
Oh my, I used it as a variable. Reading documentation again and checking it's usage - at first glance it seems the problem could be with a need to efficiently retrieve all subdirectories before enumerating and reading BUILDSYSTEM_TARGETS value from each of them and than setting targets properties - researching it now.

janvorli replied:
That's why I've said in one of my previous responses that we should use it at the end of the root CMakeLists.txt. At that point, all subdirectories were already processed.

4creators replied:
Yeah, I have noticed that in your comment and even did that when tested usage of this property -
 nevertheless I have used it as a variable .... not directory property.

weshaggard replied:
@4creators as an alternative were you able to find any way to have CMake import a custom msbuild targets/props file into all projects? If so then we could put whatever global properties we want in there and just always have it imported.

4creators replied:
@weshaggard 
It is my favorite solution to the problem but unfortunately it is impossible according to my knowledge - I have described this in PR description:

> Researching CMake support for generation of custom properties/nodes in MSBuild
project files I have found to viable implementation options:

> use VS_GLOBAL_

> use add_compile_options(/MPx) where x is expected number of CL processess

This why I have suggested to create such feature in CMake and commit it to upstream product see https://github.com/dotnet/coreclr/pull/14578#discussion_r145780476

weshaggard replied:
Given we plan to move to MSBuild15 only soon another option is to bypass CMake for this matter completely and just use an msbuild extension point. See https://github.com/Microsoft/msbuild/blob/master/src/Tasks/Microsoft.Common.props#L36 which will import the closest Directory.Build.props file that if finds up the directory tree. Given these projects are generated under the bin\ directory you should be able to put one in the root of the coreclr repo and it will automatically get imported. This should give you the flexibility to set the properties you need without making any CMake changes.

4creators commented:
@dotnet-bot test Windows_NT x64 Debug Build and Test

4creators commented:
PTAL
@janvorli @weshaggard 
I have implemented BUILDSYSTEM_TARGETS based centralized method but have not to use it due to 3 reasons:
1. We need to add every library, executable, custom target through centralized methods since they contain logic which excludes targets from cross component builds unless they are needed. Otherwise targets may slip into installations even when not expected there.
2. BUILDSYSTEM_TARGETS usage requires first gathering of all directory names through enumeration of repo source tree and then enumeration of all build targets for each directory.
3. CMake 3.7 is not available in official Ubuntu 14.04 trusty repository and it would be necessary to use PPAs if we ever wanted to use BUILDSYSTEM_TARGETS in Unix build see: https://packages.ubuntu.com/trusty/cmake3
4. It is possible to efficiently create global property containing all created targets due to 1. and it can be used both for setting VS_GLOBAL_CL_MPCount property and book keeping purposes even on Unixes with CMake 2.8.12.

If CI would be green I would prefer to merge PR now is it will fix already two problems which I have introduced: #14543 and #14605

Remaining mainly cosmetic work is tracked by: #14596

weshaggard commented:
@4creators did get a chance to explore my Directory.Build.props approach? https://github.com/dotnet/coreclr/pull/14578#discussion_r145852999

4creators commented:
@weshaggard Not yet, but my understanding is that it can be used in future when all coreclr will move MSBuild 15 - or is it possible already now?

weshaggard commented:
>@weshaggard Not yet, but my understanding is that it can be used in future when all coreclr will move to MSBuild 15 - or is it possible already now?

I think it should be possible now as all the infra already depends on MSbuild15 and given this is just an optimization if some folks still use the older version for now they just won't get this benefit which seems like a reasonable thing for now.

4creators commented:
OK will try to use this approach and your are right that it is the best solution. It may even help others to make faster decision to move to MSBuild 15.

4creators commented:
Added support for MSBuild 15 extension point which is used when VS 15.0 or higher is detected. Rearranged `run build` command line parameters to allow for general pass through syntax after -- delimiter. CI is green.

@weshaggard @janvorli 
PTAL

weshaggard commented on functions.cmake:
```diff
@@ -232,7 +233,13 @@ function(_add_library)
+    set_property(GLOBAL APPEND_STRING PROPERTY CLR_CMAKE_CORECLR_TARGETS "${ARGV0};")  
+endfunction()
+
+function(_add_custom_target)
```
Why do we still need to cmake changes once you switched to Directory.Build.props?

4creators replied:
There are 3 reasons I have decided it is worth to keep it:

1. Centralizing CMake logic for target addition (library, executable, custom target) allows to detect build logic bugs which otherwise would be missed - see comment https://github.com/dotnet/coreclr/pull/14578#issuecomment-337929416 on unneeded superpmi assemblies being added to cross components binaries bug which is tracked by issue #14596.

2. Allows to continue supporting VS 2015 (MSBuild 14) based builds with current CMake based method (VS_GLOBAL_CL_MPCount property).

3. Allows to prevent picking up by MSBuild extension point any other `Directory.Build.props` file by limiting overriding of `DirectoryBuildPropsPath` with auto discovered files in directory tree below project root. It is achieved by injecting `DirectoryBuildPropsPath` property set to absolute path to `$(RepoRoot)/Directory.Build.props` using VS_GLOBAL_DirectoryBuildPropsPath property.

4creators replied:
I plan to write short supplement to native build dcos to describe Hot To for current CMake logic and best practises when adding new project/files.

janvorli replied:
I was hoping that we will use the BUILDSYSTEM_TARGETS and require cmake >= 3.7 instead of manually creating the property and having to modify the _add_xxxxxx functions. Is there aomething preventing us doing that?

weshaggard replied:
For (1) I agree with @janvorli if we can avoid needing to override that method then we should try to do that.

For (2) I don't feel it is necessary as VS2015 support will be removed as soon as we get all our official build and CI machines with VS2017 installed so I wouldn't worry about that reason. 

For (3) why do you feel you need to override the DirectoryBuildPropsPath? That is essentially killing the extensibility hook for anyone else that may try to use it for other purposes in other layers of the directory structure, I would avoid that if we can.

janvorli commented on CMakeLists.txt:
```diff
@@ -615,6 +628,30 @@ set(BuildToolsDir "${CLR_DIR}/packages/Microsoft.DotNet.BuildTools.CoreCLR/${Bui
 
 #------------------------------
+# Add project properties to 
+# control CL prallelism
```
A nit - prallelism -> parallelism

janvorli commented on CMakeLists.txt:
```diff
@@ -615,6 +628,30 @@ set(BuildToolsDir "${CLR_DIR}/packages/Microsoft.DotNet.BuildTools.CoreCLR/${Bui
+      )
+    endif($ENV{VisualStudioVersion} VERSION_LESS "15.0")
+  endif(WIN32)
+  message(STATUS "Created target:\t${coreclrTarget}")
```
I guess this was just your debugging message, can you please remove it?

4creators commented:
@weshaggard @janvorli 

There are some contradicting code review comments which should be resolved in order to make final corrections to PR.

1. Final decision with regard to using `BUILDSYSTEM_TARGETS` injection mechanism versus MSBuild 15 `Microsoft.Common.props` extension point.

- Usage of `BUILDSYSTEM_TARGETS` would require injecting CL_MPCount properties via VS_GLOBAL_CL_MPCount mechanism whereas usage of extension point does not require any CMake modifications.

2. Decision on using `add_library`,  `add_executable`, `add_custom_target` methods versus wrapper methods `_add_library`,  `_add_executable`, `_add_custom_target`. 

I understand that decision is in favor of not using overrides. Nevertheless it is important to look at consequences:

- majority of targets (libraries, executables) are already being added using directly or indirectly `_add_library`,  `_add_executable` wrappers. If we revert this should we revert  all 72 - 74 targets or just 4 libraries and 2 executables which were included in this PR
- **wrappers contain additional logic** which checks if target should be built for cross component compilation - by adding all targets via wrapper it was possible to catch build logic errors where libraries and executables were built and installed despite the fact that they were not needed see:
https://github.com/dotnet/coreclr/pull/14578#issuecomment-338215749,  https://github.com/dotnet/coreclr/pull/14578#issuecomment-337898588 and https://github.com/dotnet/coreclr/pull/14578#issuecomment-337929416
- separate problem is an `_add_custom_target` function which is used for 17 targets - it can be removed without any side effects

Finally there is an opportunity to collect all CMake created target references `cheaply` just by using wrappers - see https://github.com/dotnet/coreclr/pull/14578#issuecomment-338215749. This may be very useful in optimization of project build graph. Currently for around 25% of native build time  CPU usage is around 15% for 8 core machine. I assume that it would be possible to cut build time by additional 20% on top of already saved 11% by optimizing build graph execution paths unless MSBuild has already found and is using critical path.

Currently CI native build time on Linux is around 7 minutes and build time on Windows is above 10 minutes. It should be possible to have them very close to each other.

janvorli commented on CMakeLists.txt:
```diff
@@ -40,6 +40,9 @@ if (PYTHON STREQUAL "PYTHON-NOTFOUND")
 endif()
 
+# Track CMake CoreCLR targets in global property
+set_property(GLOBAL PROPERTY CLR_CMAKE_CORECLR_TARGETS "")
```
With your latest change, you don't use the property for anything - you just fill it in. So I am a bit confused about this change.

4creators replied:
I have removed everything as per @weshaggard suggestion. The property `CLR_CMAKE_CORECLR_TARGETS` is left for future use in creating optimal build graph. If it should be removed as well and added later than I will do it.

There is still an open problem of reverting usage of `_add_xxx` wrappers. The question is in my previous comment at point 2: https://github.com/dotnet/coreclr/pull/14578#issuecomment-341141762

janvorli replied:
@4creators 
> This may be very useful in optimization of project build graph.

I am not sure how we would do that. The build orchestration is fully in the hands of cmake. The difference in the build time between Windows and Linux is likely due to the difference between the performance of the compilers and the underlying build tools (make vs msbuild). 

So I think that we should remove the added property, the related changes in the _add_xxx functions and also the _add_custom_target function. As for the new usages of the _add_library and _add_executable wrappers that you've added and the related additions to the CLR_CROSS_COMPONENTS_LIST, I would keep these.

4creators replied:
OK Will do it

4creators commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

4creators commented:
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test

4creators commented:
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test
@dotnet-bot Ubuntu arm64 Cross Debug Innerloop Build 
@dotnet-bot Ubuntu armlb Cross Debug Innerloop Build 
@dotnet-bot Ubuntu16.04 armlb Cross Debug Innerloop Build

4creators commented:
@dotnet-bot test Ubuntu arm64 Cross Debug Innerloop Build
@dotnet-bot  test  Tizen armel Cross Checked Innerloop Build and Test

janvorli commented on build.cmd:
```diff
@@ -439,7 +439,7 @@ if %__BuildNative% EQU 1 (
     set __MsbuildErr=/flp2:ErrorsOnly;LogFile=!__BuildErr!
 
-    @call %__ProjectDir%\run.cmd build -Project=%__IntermediatesDir%\install.vcxproj -MsBuildLog=!__MsbuildLog! -MsBuildWrn=!__MsbuildWrn! -MsBuildErr=!__MsbuildErr! -configuration=%__BuildType% %__NativePlatformArgs% %__RunArgs% -ExtraParameters="/p:ForceImportBeforeCppTargets=%__ProjectDir%/clr.nativebuild.props /m:2" %__UnprocessedBuildArgs%
+    @call %__ProjectDir%\run.cmd build -Project=%__IntermediatesDir%\install.vcxproj -MsBuildLog=!__MsbuildLog! -MsBuildWrn=!__MsbuildWrn! -MsBuildErr=!__MsbuildErr! -configuration=%__BuildType% %nativePlatfromArgs% %__RunArgs% -MSBuildNodeCount="/m:2" %__UnprocessedBuildArgs%
```
The change from __NativePlatformArgs to nativePlatformArgs seems incorrect

4creators replied:
Yep, my bad.

4creators commented:
Addressed all code review feedback, CI is green.
PTAL @janvorli @weshaggard

weshaggard commented:
@4creators thanks for the contribution. Can you please clean-up the commits and we can get this merged?
 Or we can just squash them all away.

janvorli commented:
@weshaggard squashing seems fine, there doesn't seem to be any value preserving the individual steps
